### PR TITLE
[Product] Remove product's avaiability

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets |mentioned in #7298|
| License         | MIT |

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -168,6 +168,7 @@ After:
 
 * `ProductVariant::$name` property (and corresponding getter and setter) was removed to make it translatable. Therefore, `ProductVariantTranslation` was introduced with one `$name` property. All product variants names are migrated to new concept with migration `Version2016121415313`. Look at [this PR](https://github.com/Sylius/Sylius/pull/7091) if you have any problems with upgrade.
 * `ProductAssociationType::$name` property (and corresponding getter and setter) was removed to make it translatable. Therefore, `ProductAssociationTypeTranslation` was introduced with one `$name` property. All product association types names are migrated to new concept with migration `Version20161219160441`. Look at [this PR](https://github.com/Sylius/Sylius/pull/7134) if you have any problems with upgrade.
+* `Product:$avaiableOn` and `Product:$avaiableUntil`  properties (and corresponding getters and setters) were removed. All products are migrated to new concept with migration `Version20170208102345`. Look at [this PR]() if you have any problems with upgrade.
```
Comment (Zales0123) [inline UPGRADE.md]: There is no link to PR :(

File: docs/components/Product/basic_usage.rst [hunk 2]
```diff
@@ -12,9 +12,6 @@ Creating a product
 
    $product = new Product();
 
-   $product->getCreatedAt(); // Returns the \DateTime when it was created.
```
Comment (Zales0123) [inline docs/components/Product/basic_usage.rst]: Why removing this?

File: src/Sylius/Component/Product/Model/Product.php [hunk 3]
```diff
@@ -74,8 +64,6 @@ public function __construct()
     {
         $this->initializeTranslationsCollection();
 
-        $this->createdAt = new \DateTime();
```
Comment (Zales0123) [inline src/Sylius/Component/Product/Model/Product.php]: And this?

File: UPGRADE.md [hunk 4]
```diff
@@ -168,6 +168,7 @@ After:
 
 * `ProductVariant::$name` property (and corresponding getter and setter) was removed to make it translatable. Therefore, `ProductVariantTranslation` was introduced with one `$name` property. All product variants names are migrated to new concept with migration `Version2016121415313`. Look at [this PR](https://github.com/Sylius/Sylius/pull/7091) if you have any problems with upgrade.
 * `ProductAssociationType::$name` property (and corresponding getter and setter) was removed to make it translatable. Therefore, `ProductAssociationTypeTranslation` was introduced with one `$name` property. All product association types names are migrated to new concept with migration `Version20161219160441`. Look at [this PR](https://github.com/Sylius/Sylius/pull/7134) if you have any problems with upgrade.
+* `Product:$avaiableOn` and `Product:$avaiableUntil`  properties (and corresponding getters and setters) were removed. All products are migrated to new concept with migration `Version20170208102345`. Look at [this PR](https://github.com/Sylius/Sylius/pull/7451) if you have any problems with upgrade.
```
Comment (pamil) [inline UPGRADE.md]: Double colons needed, `Product:$foo` -> `Product::$foo`

Comment (pamil) [inline UPGRADE.md]: avaiable -> avai**l**able

File: docs/components/Product/basic_usage.rst (hunk 2 above)
Comment (tuka217) [inline docs/components/Product/basic_usage.rst]: Because product don't have this method

File: src/Sylius/Component/Product/Model/Product.php (hunk 3 above)
Comment (tuka217) [inline src/Sylius/Component/Product/Model/Product.php]: See above

File: UPGRADE.md (hunk 4 above)
Comment (pamil) [inline UPGRADE.md]: Anyway, I don't see any new concept in that migration, it just removes that fields. `to new concept` -> `to a new concept`

File: src/Sylius/Component/Product/Model/Product.php [hunk 5]
```diff
@@ -197,46 +185,6 @@ public function setMetaDescription($metaDescription)
     /**
      * {@inheritdoc}
      */
-    public function isAvailable()
-    {
-        return (new DateRange($this->availableOn, $this->availableUntil))->isInRange();
```
Comment (pamil) [inline src/Sylius/Component/Product/Model/Product.php]: Is `DateRange` still used anywhere after deleting these properties?

File: src/Sylius/Component/Product/Model/Product.php (hunk 3 above)
Comment (tuka217) [inline src/Sylius/Component/Product/Model/Product.php]: I didn't see the TimestampableTrait

File: UPGRADE.md (hunk 4 above)
Comment (GSadee) [inline UPGRADE.md]: `Product:$availableOn` and `Product:$availableUntil`  properties

Comment (venyii) [inline UPGRADE.md]: `avaiable` => `available`

Comment (tuka217) [inline UPGRADE.md]: So what should be written here?

**Comment by pjedrzejewski** [comment]
Thank you Ania!

**Comment by tchapi** [comment]
Argh — can you explain the rationale ? The discussion was closed without decision on this and now it's another BC break for almost nothing :(

**Comment by Zales0123** [comment]
@tchapi well, it's not entirely true ;) The feeling from discussion in https://github.com/Sylius/Sylius/issues/7298 was rather for removing this fields than keep it. Two main reasons were:
1. They're not used in Sylius and therefore not tested properly
2. Implementing them back (for somebody that uses them) should really be an easy-peasy.

**Comment by tchapi** [comment]
But then is there a reason the fields remain in `Product.php` ?

File: src/Sylius/Bundle/PayumBundle/Extension/UpdatePaymentStateExtension.php [hunk 6]
```diff
@@ -55,8 +55,18 @@ public function onExecute(Context $context)
...
+        
+        if ($previousStackSize > 1) {
             return;
+        } 
+        
+        if ($previousStackSize === 1) {
+            $previousActionClassName = get_class($previousStack[0]->getAction());
+            if (false === stripos($previousActionClassName, 'NotifyNullAction')) {
+                return;
+            }
```
Comment (okwinza) [inline src/Sylius/Bundle/PayumBundle/Extension/UpdatePaymentStateExtension.php]: This looks completely out of scope of this PR, no?

**Comment by okwinza** [comment]
@tchapi looks like someone just forgot about them :)

**Comment by pjedrzejewski** [comment]
@tchapi We do not support this functionality properly (no tests, no logic in repositories), so no reason to keep fields. Hopefully we can add this in future versions (cause it is popular feature). Re-adding this fields is quite easy, hm? :)
